### DISCUSSION TITLE
tutorials/yolo-object-features/

### DISCUSSION THREAD

**Discussion by giscus** [Y-T-G/Yasins-Keep#9] [Giscus]
# tutorials/yolo-object-features/

Extract the object-level features from YOLO for downstream tasks without extra overhead.

https://y-t-g.github.io/tutorials/yolo-object-features/

**Comment by Burhan-Q** [comment]
Nice guide! Wonder if it would make sense to make this function a bit more flexible on downsampling (in case the channels 

```diff
+ from math import gcd

def get_object_features(feat_list, idxs):
    # feat_list would contain feature maps in grid format (N, C, H, W), where each (H,W) is an anchor location.
    # We permute and then flatten the grid so that each grid and its feature vectors
    # correspond with the indexes of the prediction. We also downsample the vector to the smallest one (64).
+  smol = gcd([f.shape[1] for f in feat_list])  # smallest channel size; shape is [b, ch, h, w]
-   obj_feats = torch.cat([x.permute(0, 2, 3, 1).reshape(-1, 64, x.shape[1] // 64).mean(dim=-1) for x in feat_list], dim=0)
+   obj_feats = torch.cat([x.permute(0, 2, 3, 1).reshape(-1, smol, x.shape[1] // smol).mean(dim=-1) for x in feat_list], dim=0)
    return obj_feats[idxs]

# Get features of every detected objected in the final output.
obj_feats = get_object_features(result[:3], idxs[0].tolist())
```

**Reply by Y-T-G** [reply]
Ah that's clever. Yeah, that should be more flexible.

**Comment by zshureih** [comment]
This worked really well when my batch size was one. However when I tried to increase the batch size I ran into issues with non_max_suppression due to `xk` not updating properly. To solve this issue, I moved the instantiation of `xk` to within the loop that iterates through the predictions, and specify which `xc` I'd like to track with `xi`. There's probably a more elegant solution but I've confirmed that the number of indices output at the end matches the number of predicted bounding boxes for each image in the batch.

Here was my solution:
```python
 def non_max_suppression(
     prediction,
     conf_thres=0.25,
     
     ...
     
     mi = 4 + nc  # mask start index
     xc = prediction[:, 4:mi].amax(1) > conf_thres  # candidates

     # Settings
     # min_wh = 2  # (pixels) minimum box width and height
     time_limit = 2.0 + max_time_img * bs  # seconds to quit after
     multi_label &= nc > 1  # multiple labels per box (adds 0.5ms/img)

     prediction = prediction.transpose(-1, -2)  # shape(1,84,6300) to shape(1,6300,84)

     if not rotated:
         if in_place:
             prediction[..., :4] = xywh2xyxy(prediction[..., :4])  # xywh to xyxy

     ...

     t = time.time()
     output = [torch.zeros((0, 6 + nm), device=prediction.device)] * bs
+    feati = [torch.zeros((0, 1), device=prediction.device)] * bs
     for xi, x in enumerate(prediction):  # image index, image inference
         # Apply constraints
         # x[((x[:, 2:4] < min_wh) | (x[:, 2:4] > max_wh)).any(1), 4] = 0  # width-height
-        x = x[xc[xi]]  # confidence
+        # To keep track of the prediction indices that remain at the end, we create an indices
+        # list that will be applied the same filters that get applied to the original predictions.
+        # That way, at the end, we will have xk with only the indices of the predictions that
+        # have not been eliminated.
+        xk = torch.tensor([list(range(len(i))) for i in xc[xi].unsqueeze(0)], device=prediction.device).transpose(-1,-2)
+
+        filt = xc[xi]
+        x = x[filt]  # confidence
+        xk = xk[filt] # indices update

         # Cat apriori labels if autolabelling
         if labels and len(labels[xi]) and not rotated:
         
         ...
```

**Reply by Y-T-G** [reply]
Thanks for the solution. I didn't test it with larger batch sizes.

**Reply by Y-T-G** [reply]
I updated the code with the fixed version that can perform both NMS and retrieve the object features in batches. `get_result_with_features()` can now take in multiple images.

**Comment by Maelic** [comment]
Hi, thank you for this interesting tutorial!

I am working on implementing a custom Pooler (Pooling + ROIAlign) for extracting features at different scales (concatenating all 3 levels for each ROI). My approach is different as I assume you can use YOLO's features more or less like a RCNN-type network (by pooling the features at different scales).
I have modify your notebook with my code and the results seems to be better a bit better (i.e. more aligned with the Resnet features). 
You can have a look if you want:
https://colab.research.google.com/drive/1LprWKiQRog9dT7DpSFQItfUA2tDnUgOJ?usp=sharing

The core of the code is as follow:

```
def level_mapper(indices):
        idx_to_level = [4800,6000,6300]
        target_lvls = torch.zeros_like(indices, dtype=torch.int64)

        # Assign values based on conditions
        target_lvls[indices < idx_to_level[0]] = 0
        target_lvls[(indices >= idx_to_level[0]) & (indices < idx_to_level[1])] = 1
        target_lvls[indices >= idx_to_level[1]] = 2

        # Convert to int64 and adjust for zero-based indexing
        return target_lvls.to(torch.int64)

def get_object_features_with_pooler(feat_list, idxs, boxes, cat_all_levels=True):
    global_avg_pool = nn.AdaptiveAvgPool2d((1, 1)) # should be replaced by a linear layer

    scales = [(80,60),(40,30),(20,15)]
    img_size = 640
    num_rois = len(boxes)
    dtype = feat_list[0].dtype
    spatial_scales = [s[0] / img_size for s in scales]
    num_channels = 192
    cat_all_levels = cat_all_levels
    # reduce channel using interpolate
    feat_list = [x.permute(0, 2, 3, 1) for x in feat_list]

    feat_list[1] = F.interpolate(feat_list[1], size=192)
    feat_list[2] = F.interpolate(feat_list[2], size=192)
    feat_list = [x.permute(0, 3, 1, 2) for x in feat_list]


    zeros = torch.full((num_rois, 1) , 0)
    boxes = torch.cat((zeros, boxes), dim=1)

    poolers = [ROIAlign([7,7], spatial_scale=spatial_scales[0], sampling_ratio=2), ROIAlign([7,7], spatial_scale=spatial_scales[1], sampling_ratio=2), ROIAlign([7,7], spatial_scale=spatial_scales[2], sampling_ratio=2)]
    levels = level_mapper(idxs[0])
    final_channels = num_channels * 3 if cat_all_levels else num_channels
    result = torch.zeros((num_rois, final_channels, 7, 7), dtype=dtype)
    for level, (per_level_feature, pooler) in enumerate(zip(feat_list, poolers)):
      if cat_all_levels:
        result[:,level*num_channels:(level+1)*num_channels,:,:] = pooler(per_level_feature, boxes).to(dtype)
      else:
        idx_in_level = torch.nonzero(levels == level).squeeze(1)
        rois_per_level = boxes[idx_in_level]
        result[idx_in_level] = pooler(per_level_feature, rois_per_level)
    output = global_avg_pool(result) # should be replaced by a linear layer
    output = output.view(num_rois, final_channels)
    return [output]
```

Let me know what you think!

**Reply by Y-T-G** [reply]
Interesting. Yes, one of the issues is comparing the features across different scales. There could be better approaches for that. I tried mapping the grid cells from the larger scales to the approximate one from the smallest scale instead of mean reduce, but the result wasn't that better.

Do you have any comparison with your method?

**Reply by Maelic** [reply]
Yes this is what the ROIAlign algorithm is really good at, finding corresponding features for a given bbox at a given scale. I think your method is not optimal because you select a single anchor vector to represent each box, which may lack some context. I choose to implement the ROIAlign with a size of 7*7 grid cells as in the original MaskRCNN paper, but maybe a better value could be found for YOLO. Also, there are other ROIAlign variants that could be tested.

I compared your approach and mine only on a few samples, here is an example (your method is WITHOUT and mine is WITH):

![download](https://github.com/user-attachments/assets/b98e9097-da8b-43ab-a862-9b855347aadb)

You can see that your method is not very consistent, with very low values (0.14 for comparing two cars is weird). I think the difference here comes from the fact that aggregating features from different scales give a lot more contextual information.
My approach is a bit slower though (on colab at least).

This would, of course, require a more thorough investigation:)

**Reply by Y-T-G** [reply]
> Yes this is what the ROIAlign algorithm is really good at, finding corresponding features for a given bbox at a given scale. I think your method is not optimal because you select a single anchor vector to represent each box, which may lack some context.

I can see why you would think that. Especially when thinking in terms of two-stage detectors.

The thing is, the feature vector that is extracted is actually the exact vector that was by YOLO used to get the final localization and classification result for that particular object. In fact, you could run only that same exact vector through the head's classification and regression branch again to get the same final output. This would mean that it does have the necessary context about the object because otherwise the prediction, especially localization, wouldn't have been accurate.

But I guess using features from multiple scales is probably a good idea. I think this can be done without ROIAlign by getting a mapping between each of the grids, 80x80, 40x40, 20x20 and then comparing the features from the corresponding scales between objects and getting the average similarity. This is more aligned with how YOLO works. Each grid cell in YOLO is directly connected to an object, unlike in two-stage networks, which requires ROIAlign.

For example, a grid cell in 20x20 feature map maps to 2x2 cell in 40x40 and 4x4 cell in 80x80. So for each object, you could extract 3 vectors, one from each of the scales and then compare the vectors from the same scales between objects.

**Reply by Maelic** [reply]
Yes I agree with you, aggregating the features by index will be the easiest solution. But in my opinion (and with my experiments), it is difficult to relate features from different scales without aligning the corresponding box. 
This for three reasons:
1. The size of the box that corresponds to an anchor vector will change from one image to another, meaning that if you have a feature that corresponds to a box of size m at the first scale, in some images it will correspond to 3 different features at the next scale or 7 different ones in another image etc...
2. As far as I understand how yolo works, neighboring anchors likely correspond to representations of the same object because they will be decoded as close bounding boxes in the later stages (this is why you need NMS). Then, it is probably efficient to aggregate neighboring features corresponding to a single object (using ROIAlign) to obtain more information about that object.
3. It is unclear how a single feature vector from YOLO represents enough information to be used for anything else than regression and classification with YOLO, because in YOLO the backbone and heads are trained jointly. If you take YOLO s or n, the vector extracted will be only of size 64 or 128 which is likely insufficient. As a comparison, other feature extractors (clip, swin transformer, dinov2 etc) will extract at least 512 features, most of the time aggregated from multiple scales.

But all-in-all it really depends on what kind of features you want to extract, and for which task you want to use them. My task is relation prediction so I'm more looking for high-level semantic features than let's say appearance-based features. This is why I'm trying to extract as much context as possible (and from later stages of the network).

I will see if I can implement your solution to relate features by index only, and see if it's better than ROIAlign. Do you have any hint on how we could do that based on my code or yours? Likely dividing/multiplying the index won't be enough.

**Reply by Y-T-G** [reply]
1. Yes, that's a genuine problem. It tried `map_to_smallest` (see below) to make them all use the same feature map. I'm not sure if it's better.
2. Yes, the neighboring anchors should have some overlap. But there's also a difference with two-stage detectors in the sense that ROIAlign is used to train two-stage models to learn features which enforces this behavior. But for YOLO, the anchors are simply assigned to the ground truth. And for anchor point based models like YOLOv8, the anchor closest to the center of GT is prioritized.
3. It's possible the features are too sparse, especially for smaller models like you mentioned.

I tried this. `map_to_smallest` maps the indices from larger feature maps to the smallest one. This avoids having to compares across different feature maps, and the features may be more comparable. The similarity score here is low too. `map_to_scales` maps each indice to all the three feature maps and returns them. So for each object, there are 3 different vectors being extracted. The cosine similarity is calculated across the same scales and then averaged. This returns higher similarity scores. I'm not sure if all the code is correct, especially the mapping. It may possibly have some logic error I overlooked.

```python
def map_level(from_shape, to_shape, idx):
    # Maps the indice from one feature map to anoter
    ks = to_shape / from_shape   # Get the kernel size
    idx = torch.tensor(idx)
    # Returns the x_start, y_start, x_end, y_end indices to slice the feature map TO which the FROM indice was mapped
    return torch.hstack((torch.floor(idx * ks), torch.ceil((idx + 1) * ks))).int()

def get_scale_feats(idx, feat_list, batch_i):
    feat_shapes = np.array([(feat.shape[-2], feat.shape[-1]) for feat in feat_list])
    # Determine which feature map the idx belongs to
    cum_sizes = np.cumsum([s.prod() for s in feat_shapes])
    if idx < cum_sizes[0]:
        _idx, from_shape = idx, feat_shapes[0]
    elif idx < cum_sizes[1]:
        _idx, from_shape = idx - cum_sizes[0], feat_shapes[1]
    else:
        _idx, from_shape = idx - cum_sizes[1], feat_shapes[2]

    # Compute (i, j) coordinates
    i, j = _idx % from_shape[1], _idx // from_shape[1]

    # Get corresponding regions from all feature maps
    slices = [map_level(from_shape, to_shape, (i, j)) for to_shape in feat_shapes]

    # Extract feature patches
    obj_feat = [
        feat[batch_i][:, y1:y2, x1:x2].flatten(-2).mean(dim=-1)
        for (x1, y1, x2, y2), feat in zip(slices, feat_list)
    ]

    return obj_feat

def map_to_feat(idx, feat_list):
    # Map to smallest feature map
    feat_shapes = np.array([(feat.shape[-2], feat.shape[-1]) for feat in feat_list])
    # Determine which feature map the idx belongs to
    cum_sizes = np.cumsum([s.prod() for s in feat_shapes])
    if idx < cum_sizes[0]:
        _idx, from_shape = idx, feat_shapes[0]
    elif idx < cum_sizes[1]:
        _idx, from_shape = idx - cum_sizes[0], feat_shapes[1]
    else:
        _idx, from_shape = idx - cum_sizes[1], feat_shapes[2]

    # Compute (i, j) coordinates
    i, j = _idx % from_shape[1], _idx // from_shape[1]
    
    # Map to last feature map, i.e. smallest
    mapped = map_level(from_shape, feat_shapes[-1], (i, j))
    return (mapped[1] * feat_shapes[-1][1]) + mapped[0]

def get_object_features(feat_list, idxs, method = "downsample"):
    # feat_list would contain feature maps in grid format (N, C, H, W), where each (H,W) is an anchor location.
    if method == "map_to_smallest":
        # the feature map sizes are different, so we map a grid cell from a larger feature map to a grid cell in the smallest feature map (20x20)
        _idxs = [torch.stack([map_to_feat(i, feat_list) for i in idx]) for idx in idxs]
        obj_feats = torch.cat([x.permute(0, 2, 3, 1).flatten(1,2) for x in feat_list[-1:]], dim=1)
        return [feats[idx] for feats, idx in zip(obj_feats, _idxs)] # for each image in batch, indexed separately
    elif method == "map_scales":
        # the feature map sizes are different, so we get the feature from each scale for all the objects
        # idxs is batch of idxs for each image; feat_list contains batched features (N, C, H, W)
        return [[get_scale_feats(idx, feat_list, batch_i) for idx in batch_idx] for batch_i, batch_idx in enumerate(idxs)]
    else:
        # We permute and then flatten the grid so that each grid and its feature vectors
        # correspond with the indexes of the prediction. We also downsample the vector to the smallest one (64).
        s = gcd(*[x.shape[1] for x in feat_list]) # smallest vector length (64 for YOLOv8)
        obj_feats = torch.cat([x.permute(0, 2, 3, 1).reshape(x.shape[0], -1, s, x.shape[1] // s).mean(dim=-1) for x in feat_list], dim=1)
        return [feats[idx] for feats, idx in zip(obj_feats, idxs)] # for each image in batch, indexed separately

# map_scales compares features from each scale separately
method = "map_scales"
# Get features of every detected objected in the final output.
obj_feats = get_object_features(result[:3], idxs, method=method)

# Similarity between first and second object
# The order corresponds to the order of boxes in results below
image_i = 0  # index of the image in batch
obj1_i = 1
obj2_i = 7
if method == "map_scales":
  # Each scale is compared separately
  print(torch.hstack([cosine_similarity(feat1, feat2, dim=0) for feat1, feat2 in zip(obj_feats[image_i][obj1_i], obj_feats[image_i][obj2_i])]).mean())
else:
  print(cosine_similarity(obj_feats[image_i][obj1_i], obj_feats[image_i][obj2_i], dim=0))
```

I observed that similarity score calculation across the larger/earlier feature maps have larger values, while the later ones return smaller values.

**Reply by Maelic** [reply]
Thank you for your code, I will try it out!

Also, have you tried to use features from other layers of the model such as features from layers 4,6,9 which correspond to the end of the backbone? I have seen some weird stuff in features from layers 15,18,21 when visualizing the features. It seems that the PAN-FPN is actually learning some box-related information (or at least it is my interpretation), see the following example:

![image](https://github.com/user-attachments/assets/def9c4c1-fa61-41e5-bcfb-4deec80a191b)

Here we can see some features extracted from layer 15, some contain weird vertical and horizontal lines (see top right or left). I know most of the horizontal lines are due to re-shaping the image to a square size (I am using 640*640 input size here) but some other lines seem to already encode some box-related information around the centered object (the train in that case).

By comparison, features from the layer 4 are more looking like features we get from traditional resnet-based feature extractors:

![image](https://github.com/user-attachments/assets/749a1fe9-3f57-4150-9d3f-97bfa0437cee)

So, using layers 4,6,9 may lead to better similarity using your method.

The original image for reference:

![download](https://github.com/user-attachments/assets/6a13ebc2-5107-4b4e-bbee-20bdd6f4f411)

**Comment by huaiguang233** [comment]
Thank you for providing such a detailed idea. However, I encountered an issue when using this approach to obtain Object Level Features in YOLOv10. I think it's because v10 is anchor free model. So how can I obtain Object Level Features in an anchor free architecture, like YOLOv10?

**Reply by Y-T-G** [reply]
Instead of NMS, for YOLOv10 you need to get the indices from [here](https://github.com/ultralytics/ultralytics/blob/dbe2b5d4bb27803e150a7e5fea4b5d4449032690/ultralytics/nn/modules/head.py#L181) and then use that to get the features.

**Comment by rromanchuk** [comment]
It would be really really  nice if someone took these learnings and was able to implement an interface for
https://docs.voxel51.com/api/fiftyone.core.models.html#fiftyone.core.models.EmbeddingsMixin 

There is already a nice integration https://docs.voxel51.com/integrations/ultralytics.html?highlight=ultralytics but obviously doesn't expose embeddings https://docs.voxel51.com/model_zoo/models.html#yolo11n-coco-torch
